api/auth/azure/azure.go

```diff
@@ -0,0 +1,204 @@
+	}
+
+	loginData := map[string]interface{}{
+		"role":                a.roleName,
+		"jwt":                 jwtResp,
+		"vm_name":             metadataRespJSON.Compute.VMName,
+		"vmss_name":           metadataRespJSON.Compute.VMScaleSetName,
+		"subscription_id":     metadataRespJSON.Compute.SubscriptionID,
+		"resource_group_name": metadataRespJSON.Compute.ResourceGroupName,
+	}
+
+	path := fmt.Sprintf("auth/%s/login", a.mountPath)
+	resp, err := client.Logical().Write(path, loginData)
+	if err != nil {
+		return nil, fmt.Errorf("unable to log in with Azure auth: %w", err)
+	}
+
+	return resp, nil
+}
+
+func WithMountPath(mountPath string) LoginOption {
+	return func(a *AzureAuth) error {
+		a.mountPath = mountPath
+		return nil
+	}
+}
+
+func WithResourceURL(url string) LoginOption {
+	return func(a *AzureAuth) error {
+		a.resource = url
+		return nil
+	}
+}
+
+// Retrieves an access token from Managed Identities for Azure Resources
+//
+// Learn more here: https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token
+func (a *AzureAuth) getJWT() (string, error) {
+	identityEndpoint, err := url.Parse(fmt.Sprintf("%s/metadata/identity/oauth2/token", metadataEndpoint))
+	if err != nil {
+		return "", fmt.Errorf("error creating metadata URL: %w", err)
+	}
+
+	identityParameters := identityEndpoint.Query()
+	identityParameters.Add(apiVersionQueryParam, metadataAPIVersion)
+	identityParameters.Add(resourceQueryParam, a.resource)
+	identityEndpoint.RawQuery = identityParameters.Encode()
+
+	req, err := http.NewRequest("GET", identityEndpoint.String(), nil)
+	if err != nil {
+		return "", fmt.Errorf("error creating HTTP request: %w", err)
+	}
+	req.Header.Add("Metadata", "true")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("error calling Azure token endpoint: %w", err)
+	}
+
+	responseBytes, err := ioutil.ReadAll(resp.Body)
+	defer resp.Body.Close()
+	if err != nil {
+		return "", fmt.Errorf("error reading response body from Azure token endpoint: %w", err)
+	}
+
+	var r responseJSON
+	err = json.Unmarshal(responseBytes, &r)
+	if err != nil {
+		return "", fmt.Errorf("error unmarshaling response from Azure token endpoint: %w", err)
+	}
+
+	return r.AccessToken, nil
+}
+
+func getMetadata() (metadataJSON, error) {
+	metadataEndpoint, err := url.Parse(fmt.Sprintf("%s/metadata/instance", metadataEndpoint))
+	if err != nil {
+		fmt.Println("Error creating URL: ", err)
+		return metadataJSON{}, err
+	}
+
+	metadataParameters := metadataEndpoint.Query()
+	metadataParameters.Add(apiVersionQueryParam, metadataAPIVersion)
+	metadataEndpoint.RawQuery = metadataParameters.Encode()
+	req, err := http.NewRequest("GET", metadataEndpoint.String(), nil)
+	if err != nil {
+		return metadataJSON{}, fmt.Errorf("error creating HTTP Request for metadata endpoint: %w", err)
+	}
+	req.Header.Add("Metadata", "true")
+
+	client := &http.Client{}
```

Comment:
Using a default http client omits setting sensible timeouts which can lead to problems if the socket is never closed, I suggest adding a timeout of e.g. 10s for all http clients

Reply:
Thanks, nice catch!